PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/configuration/ui/AzureFunctionsConfigurationPanel.kt

```diff
@@ -211,58 +208,69 @@ class AzureFunctionsConfigurationPanel: AzureRiderAbstractConfigurablePanel {
         }
 
-        noteRow(message("settings.app_services.function_app.core_tools.description"))
-
         row {
-            cell(isFullWidth = true) {
-                scrollPane(coreToolsEditor).noGrowY()
-            }
+            scrollCell(coreToolsEditor).onApply {
+                    AzureRiderSettings.setAzureCoreToolsPathEntries(properties, coreToolsEditorModel.items)
+            }.align(AlignX.FILL)
         }
 
         if (isCoreToolsFeedEnabled) {
-            row(message("settings.app_services.function_app.core_tools.download_path")) { }
             row {
-                cell(isFullWidth = true) {
-                    component(coreToolsDownloadPathEditor)
-                }
+                text(message("settings.app_services.function_app.core_tools.download_path"))
+            }
+            row {
+                var value = FileUtil.toSystemDependentName(properties.getValue(AzureRiderSettings.PROPERTY_FUNCTIONS_CORETOOLS_DOWNLOAD_PATH)
+                        .orWhenNullOrEmpty(AzureRiderSettings.VALUE_FUNCTIONS_CORETOOLS_DOWNLOAD_PATH))
+
+                textFieldWithBrowseButton(
+                        message("settings.app_services.function_app.core_tools.download_path_description"),
+                        null,
+                        FileChooserDescriptorFactory.createSingleFolderDescriptor()
+                )
+                    .bindText({ value }, { value = FileUtil.toSystemIndependentName(it) })
+                    .onApply { properties.setValue(AzureRiderSettings.PROPERTY_FUNCTIONS_CORETOOLS_DOWNLOAD_PATH, value) }
+                    .validationOnInput { validationForPath(it) }
+                    .align(AlignX.FILL)
             }
         }
 
         row {
-            placeholder().constraints(growY, pushY)
+            placeholder()
         }
     }
 
-    override fun doOKAction() {
+    private fun validationForPath(textField: TextFieldWithBrowseButton) =
+            if (!textField.text.isNullOrEmpty() && !File(textField.text).exists()) {
+                ValidationInfo(message("settings.app_services.function_app.core_tools.download_path.invalid"), textField)
+            } else {
+                null
+            }
 
-        AzureRiderSettings.setAzureCoreToolsPathEntries(properties, coreToolsEditorModel.items)
+    override val panel = createPanel().apply {
+        registerValidators(disposable)
+        reset()
+    }
 
-        if (coreToolsDownloadPathEditor.text != "" && File(coreToolsDownloadPathEditor.text).exists()) {
-            properties.setValue(
-                    AzureRiderSettings.PROPERTY_FUNCTIONS_CORETOOLS_DOWNLOAD_PATH,
-                    FileUtil.toSystemIndependentName(coreToolsDownloadPathEditor.text))
-        }
+    override fun isModified() = panel.isModified()
+
+    override fun doResetAction() = panel.reset()
+
+    override fun doOKAction() {
+        panel.apply()
+        Disposer.dispose(disposable)
```

Comment:
_(Minor)_ Is this required? I thought `dispose` will be called anyway after `doOKAction`.

If it is, then `private val disposable` is not needed, and you can just use `this` everywhere instead of `disposable`.

Reply:
Reworked it a bit.